changes/2025-09-12_introduce-metrics-interface/change.md

```diff
@@ -0,0 +1,169 @@
+
+1. MUST be simple.
+1. MUST be extensible.
+
+The following is documented to signify its importance
+even though the interface is not able to make this guarantee.
+Every implementation of the proposed interface must
+ensure the following.
+
+1. MUST NOT block the application's execution thread.
+
+## Points of Integration
+
+To collect metrics across CT's library stack multiple points of integration
+are needed in order to collect metrics across CT's libraries.
+Generally, CT's libraries work as follows:
+
+_Note: Not every Client supports the Material Provider Library.
+The Diagram below assumes it to help the mental model._
+
+```mermaid
+sequenceDiagram
+    participant Client
+    box Material Provider Library
+      participant CMM
+      participant Keyring
+    end
+    loop Content Encryption
+        Client->>Client: Content Encryption
+    end
+    Client<<->>CMM: GetEncryption/Decryption Materials
+    CMM<<->>Keyring: OnEncrypt/OnDecrypt
+```
+
+To optionally emit metrics from a top level client,
+all customer facing APIs MUST be changed to optionally accept
+a Metrics Agent. This will allow customers to define and supply one top
+level Metrics Agent; this agent will get plumbed throughout CT's stack.
+
+For example, in the ESDK for Java this would look like:
+
+```java
+final AwsCrypto crypto = AwsCrypto.builder().build();
+// Create a Keyring
+final MaterialProviders matProv =
+  MaterialProviders.builder()
+      .MaterialProvidersConfig(MaterialProvidersConfig.builder().build())
+      .build();
+
+final IKeyring rawAesKeyring = matProv.CreateRawAesKeyring(keyringInput);
+final Map<String, String> encryptionContext =
+    Collections.singletonMap("ExampleContextKey", "ExampleContextValue");
+
+// Create a Metrics Agent
+final IMetricsAgent metrics = matProv.CreateSimpleMetricsAgent(metricsAgentInput);
+// 4. Encrypt the data
+final CryptoResult<byte[], ?> encryptResult =
+    crypto.encryptData(rawAesKeyring, metrics, EXAMPLE_DATA, encryptionContext);
+final byte[] ciphertext = encryptResult.getResult();
+```
+
+This change will allow Crypto Tools to introduce a Metrics Agent in a
+non-breaking way as the Metrics Agent will be an optional parameter
+at customer facing API call sites.
+
+Currently, the ESDK client APIs models are defined [here](https://github.com/aws/aws-encryption-sdk/blob/mainline/AwsEncryptionSDK/dafny/AwsEncryptionSdk/Model/esdk.smithy#L60-L126).
+This change would see that the client APIs be changed as follows:
+
+```diff
+structure EncryptInput {
+  @required
+  plaintext: Blob,
+
+  encryptionContext: aws.cryptography.materialProviders#EncryptionContext,
+
+  // One of keyring or CMM are required
+  materialsManager: aws.cryptography.materialProviders#CryptographicMaterialsManagerReference,
+  keyring: aws.cryptography.materialProviders#KeyringReference,
+
+  algorithmSuiteId: aws.cryptography.materialProviders#ESDKAlgorithmSuiteId,
+
+  frameLength: FrameLength,
+
```

Comment:
I guess this probably isn't possible in Smithy/Smithy-Dafny but we might want to consider having a more open-ended "request override" option for things like this. (This is the same way the SDK does it.)